[MSHARED-1066] - Upgrade Plexus Archiver to 4.3.0, Improve the Reproducible Builds methods

jorsol opened the pull request:
Fixes [MSHARED-1066](https://issues.apache.org/jira/browse/MSHARED-1066) and [MSHARED-1067](https://issues.apache.org/jira/browse/MSHARED-1067)

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MSHARED) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[MSHARED-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MSHARED-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [X] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [X] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

michael-o commented on src/main/java/org/apache/maven/archiver/MavenArchiver.java:
```diff
@@ -812,28 +814,70 @@ public void setBuildJdkSpecDefaultEntry( boolean buildJdkSpecDefaultEntry )
+        }
+
+        // Number representing seconds since the epoch
         if ( StringUtils.isNumeric( outputTimestamp ) && StringUtils.isNotEmpty( outputTimestamp ) )
```
Interesting, can it be empty if it is a numeric?

jorsol replied:
Yeah, that shocked me as well, but that utility method doesn't handle empty values and return true for an empty value.

michael-o replied:
WTF? Joder!

cstamas commented:
Can we split out the two thing here? Update p-archiver (and plexus-io) is unrelated to repro methods, is it?

jorsol commented:
> Can we split out the two thing here? Update p-archiver (and plexus-io) is unrelated to repro methods, is it?

Removed plexus-io update as is actually unrelated (will open a follow-up PR).

The update of plexus-archiver and reproducible methods are actually related, the reproducible methods depends on plexus-archiver 4.3.0 as plexus-archiver also contains changes to methods. It's split in two commits, and to make them two different PRs, the plexus-archiver update should be merged first and then the reproducible methods later, yet since the dependency update is a really small change I would like to keep them together.

cstamas commented:
Kk, but it would be good to sort out maven-archiver changes, make a release:
https://issues.apache.org/jira/issues/?jql=project%20%3D%20MSHARED%20AND%20fixVersion%20%3D%20maven-archiver-3.6.0

As MASSEMBLY is eagerly  waiting updates
https://issues.apache.org/jira/issues/?jql=project%20%3D%20MASSEMBLY%20AND%20fixVersion%20%3D%203.4.0

jorsol commented:
> Kk, but it would be good to sort out maven-archiver changes, make a release **soon**: https://issues.apache.org/jira/issues/?jql=project%20%3D%20MSHARED%20AND%20fixVersion%20%3D%20maven-archiver-3.6.0

This should be ready for review, after merging this and https://github.com/apache/maven-archiver/pull/24. After that, a release can be made.

cstamas commented:
wazzup here?

michael-o commented on src/main/java/org/apache/maven/archiver/MavenArchiver.java:
```diff
@@ -97,6 +101,10 @@
 
+    private static final ZonedDateTime DATE_MIN = ZonedDateTime.parse( "1980-01-01T00:00:02Z" );
+
+    private static final ZonedDateTime DATE_MAX = ZonedDateTime.parse( "2099-12-31T23:59:59Z" );
```
Isn't this for Zip files only?

michael-o replied:
Yes, you are right. `MavenArchiver` is for JAR-derived file types only.

michael-o commented on src/main/java/org/apache/maven/archiver/MavenArchiver.java:
```diff
@@ -812,28 +820,78 @@ public void setBuildJdkSpecDefaultEntry( boolean buildJdkSpecDefaultEntry )
-            return df.parse( outputTimestamp );
+            // Parse the date in UTC such as '2011-12-03T10:15:30Z' or with an offset '2019-10-05T20:37:42+06:00'.
+            final ZonedDateTime date = ZonedDateTime.parse( outputTimestamp )
+                .withZoneSameInstant( ZoneOffset.UTC ).truncatedTo( ChronoUnit.SECONDS );
```
That reads beautifully...

michael-o commented on src/main/java/org/apache/maven/archiver/MavenArchiver.java:
```diff
@@ -812,28 +820,78 @@ public void setBuildJdkSpecDefaultEntry( boolean buildJdkSpecDefaultEntry )
+     * @throws IllegalArgumentException if the outputTimestamp is neither ISO 8601 nor an integer
+     */
+    public static Optional<Instant> parseBuildOutputTimestamp( String outputTimestamp )
+    {
```
I have the feeling that such parsing is done over and over again in many components. We need to consider to have ths at as central place.

jorsol replied:
I'm not sure what this means, are you referring to moving this out of Maven Archiver?

michael-o commented on src/test/java/org/apache/maven/archiver/MavenArchiverTest.java:
```diff
@@ -1444,54 +1447,76 @@ public void testParseOutputTimestamp()
+    @CsvSource( { "0,0", "1,1", "9,9", "1570300662,1570300662", "2147483648,2147483648",
+        "2019-10-05T18:37:42Z,1570300662", "2019-10-05T20:37:42+02:00,1570300662",
+        "2019-10-05T16:37:42-02:00,1570300662", "1988-02-22T15:23:47.76598Z,572541827",
+        "2011-12-03T10:15:30+01:00[Europe/Paris],1322903730",
```
`2011-12-03T10:15:30+01:00[Europe/Paris]` is invalid. Should not pass.

jorsol replied:
Sure, changed to OffsetDateTime to not accept the Zone.

michael-o commented on src/test/java/org/apache/maven/archiver/MavenArchiverTest.java:
```diff
@@ -1444,54 +1447,76 @@ public void testParseOutputTimestamp()
+
+    @ParameterizedTest
+    @ValueSource( strings = { "1980-01-01T00:00:01Z", "2100-01-01T00:00Z", "2100-02-28T23:59:59Z",
+        "2099-12-31T23:59:59-01:00", "1980-01-01T00:15:35+01:00[Europe/Madrid]" } )
```
Same for the last.

michael-o commented on src/test/java/org/apache/maven/archiver/MavenArchiverTest.java:
```diff
@@ -1444,54 +1447,76 @@ public void testParseOutputTimestamp()
+
+    @ParameterizedTest
+    @CsvSource( { "2011-12-03T10:15:30+01[Europe/Paris],1322903730", "2019-10-05T20:37:42+02,1570300662",
+        "2011-12-03T10:15:30+06[America/Managua],1322885730", "1988-02-22T20:37:42+06,572539062" } )
```
Here too.

michael-o commented on src/main/java/org/apache/maven/archiver/MavenArchiver.java:
```diff
@@ -812,28 +820,78 @@ public void setBuildJdkSpecDefaultEntry( boolean buildJdkSpecDefaultEntry )
      *         character
      * @since 3.5.0
-     * @throws java.lang.IllegalArgumentException if the outputTimestamp is neither ISO 8601 nor an integer
+     * @throws IllegalArgumentException if the outputTimestamp is neither ISO 8601 nor an integer
```
Does not document the out of range

michael-o commented on src/main/java/org/apache/maven/archiver/MavenArchiver.java:
```diff
@@ -812,28 +820,78 @@ public void setBuildJdkSpecDefaultEntry( boolean buildJdkSpecDefaultEntry )
+     * @return the parsed timestamp as an {@code Optional<Instant>}, {@code empty} if input is {@code null} or input
+     *         contains only 1 character (not a number)
+     * @since 3.6.0
+     * @throws IllegalArgumentException if the outputTimestamp is neither ISO 8601 nor an integer
```
Same here for the range

michael-o commented on src/test/java/org/apache/maven/archiver/MavenArchiverTest.java:
```diff
@@ -1444,54 +1447,75 @@ public void testParseOutputTimestamp()
+    }
+
+    @ParameterizedTest
+    @CsvSource( { "0,0", "1,1", "9,9", "1570300662,1570300662", "2147483648,2147483648",
```
That source is quite insane :-D

michael-o commented on src/test/java/org/apache/maven/archiver/MavenArchiverTest.java:
```diff
@@ -1444,54 +1447,75 @@ public void testParseOutputTimestamp()
+    @ParameterizedTest
+    @ValueSource( strings = { "1980-01-01T00:00:01Z", "2100-01-01T00:00Z", "2100-02-28T23:59:59Z",
+        "2099-12-31T23:59:59-01:00", "1980-01-01T00:15:35+01:00" } )
+    public void testThrownParseOutputTimestampValidRange( String outputTimestamp )
```
`testThrownParseOutputTimestampinvalidRange`, no?

michael-o commented on src/main/java/org/apache/maven/archiver/MavenArchiver.java:
```diff
@@ -812,28 +820,80 @@ public void setBuildJdkSpecDefaultEntry( boolean buildJdkSpecDefaultEntry )
      *         character
      * @since 3.5.0
-     * @throws java.lang.IllegalArgumentException if the outputTimestamp is neither ISO 8601 nor an integer
+     * @throws IllegalArgumentException if the outputTimestamp is neither ISO-8601 nor an integer, or it's not within
```
ISO 8601 (no hyphen)

michael-o commented on src/main/java/org/apache/maven/archiver/MavenArchiver.java:
```diff
@@ -812,28 +820,80 @@ public void setBuildJdkSpecDefaultEntry( boolean buildJdkSpecDefaultEntry )
+     * @return the parsed timestamp as an {@code Optional<Instant>}, {@code empty} if input is {@code null} or input
+     *         contains only 1 character (not a number)
+     * @since 3.6.0
+     * @throws IllegalArgumentException if the outputTimestamp is neither ISO-8601 nor an integer, or it's not within
```
ISO 8601 (no hyphen)

michael-o commented:
Except for the remaining comments this looks good. I will run this snapshot against MJAR, MWAR, and likely other projects just to make sure.

michael-o commented:
Please have a look (maven-source-plugin):
```
[INFO] Building: reproducible/pom.xml
[INFO] run post-build script verify.groovy
[INFO]   The post-build script did not succeed. assert buf.startsWith( "reproducible-1.0-sources.jar sha1 = 3a3687b063cfc164fbbccd1b9573b4232f540e8a" )
       |   |
       |   false
       reproducible-1.0-sources.jar sha1 = f159379802c1f0dc1083af21352286b09d364519

       encoding: UTF8
       timezone offset (minutes): -120
       M size (cmp) crc      java time     date       time           zip time   mode   name -comment; extra
       0    0 (  0)        0 1566404932000 2019-08-21 18:28:52 +0200 1326814106  40755 META-INF/ ; 0
       8   25 ( 27) ee027fb2 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 META-INF/MANIFEST.MF ; 0
       0    0 (  0)        0 1566404932000 2019-08-21 18:28:52 +0200 1326814106  40755 dir-A/ ; 0
       0    0 (  0)        0 1566404932000 2019-08-21 18:28:52 +0200 1326814106  40755 dir-C/ ; 0
       0    0 (  0)        0 1566404932000 2019-08-21 18:28:52 +0200 1326814106  40755 dir-b/ ; 0
       0    0 (  0)        0 1566404932000 2019-08-21 18:28:52 +0200 1326814106  40755 dir-b/B2/ ; 0
       0    0 (  0)        0 1566404932000 2019-08-21 18:28:52 +0200 1326814106  40755 dir-b/B4/ ; 0
       0    0 (  0)        0 1566404932000 2019-08-21 18:28:52 +0200 1326814106  40755 dir-b/b1/ ; 0
       0    0 (  0)        0 1566404932000 2019-08-21 18:28:52 +0200 1326814106  40755 dir-b/b3/ ; 0
       0    0 (  0)        0 1566404932000 2019-08-21 18:28:52 +0200 1326814106  40755 dir-d/ ; 0
       8  281 (118) d793ddf2 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 META-INF/DEPENDENCIES ; 0
       8 11358 (3949) 86e2b4b4 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 META-INF/LICENSE ; 0
       8  183 (129) aeabb55c 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 META-INF/NOTICE ; 0
       8    9 ( 11) 1c933a72 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 Uppercase.txt ; 0
       8    2 (  4) 606c3d3b 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 dir-A/A2.txt ; 0
       8    2 (  4) 890f980e 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 dir-A/A4.txt ; 0
       8    2 (  4) 6ce14823 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 dir-A/a1.txt ; 0
       8    2 (  4) 82ef290f 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 dir-A/a3.txt ; 0
       8    1 (  3) 3dd7ffa7 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 dir-C/C.txt ; 0
       8    2 (  4) 4b416ef8 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 dir-b/B2/B2.txt ; 0
       8    2 (  4) a222cbcd 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 dir-b/B4/B4.txt ; 0
       8    1 (  3) 71beeff9 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 dir-b/b.txt ; 0
       8    2 (  4) 47cc1be0 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 dir-b/b1/b1.txt ; 0
       8    2 (  4) a9c27acc 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 dir-b/b3/b3.txt ; 0
       8    1 (  3) 98dd4acc 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 dir-d/d.txt ; 0
       8   10 ( 12) d68eda01 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 executable.txt ; 0
       8    9 ( 11) fe3d51f2 1566404932000 2019-08-21 18:28:52 +0200 1326814106 100644 lowercase.txt ; 0
[INFO]           reproducible/pom.xml ............................. FAILED (4.9 s)
```

jorsol commented:
> Please have a look (maven-javadoc-plugin):

The `maven-javadoc-plugin` declares the `plexus-archiver` dependency, so it needs to be updated to 4.3.0 (or removed to be used transitively) at the same time as `maven-archiver`.

https://github.com/apache/maven-javadoc-plugin/blob/b92897088f6fcbc5ec4f3346de591eb97a1c580f/pom.xml#L288-L292

michael-o commented:
Users: https://github.com/search?q=org%3Aapache+maven-archiver+path%3A%2F+language%3A%22Maven+POM%22+language%3A%22Maven+POM%22&type=Code&ref=advsearch&l=Maven+POM&l=Maven+POM

michael-o commented:
> > Please have a look (maven-javadoc-plugin):
> 
> The `maven-javadoc-plugin` declares the `plexus-archiver` dependency, so it needs to be updated to 4.3.0 (or removed to be used transitively) at the same time as `maven-archiver`.
> 
> https://github.com/apache/maven-javadoc-plugin/blob/b92897088f6fcbc5ec4f3346de591eb97a1c580f/pom.xml#L288-L292

Done. Issue resolved.

michael-o commented:
I can confirm that maven-source-plugin passes from master, but not this PR.

jorsol commented:
> I can confirm that maven-source-plugin passes from master, but not this PR.

The IT from `maven-source-plugin` uses a sha1 checksum (which does not necessarily test that is reproducible).

And actually, the root cause is not even in this PR, you can make that test fail by just updating plexus-archiver to 4.2.7 (right now is at 4.2.4), plexus-archiver 4.2.7 fixes the order of META-INF/ and META-INF/MANIFEST.MF entries in a JAR file, this "reorder" of entries makes that the output (and the checksum) change.

So, essentially the "fix" is just to use the new sha1 checksum.

michael-o commented:
> 



> > I can confirm that maven-source-plugin passes from master, but not this PR.
> 
> The IT from `maven-source-plugin` uses a sha1 checksum (which does not necessarily test that is reproducible).
> 
> And actually, the root cause is not even in this PR, you can make that test fail by just updating plexus-archiver to 4.2.7 (right now is at 4.2.4), plexus-archiver 4.2.7 fixes the order of META-INF/ and META-INF/MANIFEST.MF entries in a JAR file, this "reorder" of entries makes that the output (and the checksum) change.
> 
> So, essentially the "fix" is just to use the new sha1 checksum.

Right, I did the fix for the order because it was broken for quite some time. Let me double check.

michael-o commented:
I will use diffoscope.

michael-o commented:
Here it is:
```
osipovmi@deblndw011x:~/var/Projekte/maven-source-plugin (master *=)
$ ~/.local/bin/diffoscope /tmp/reproducible-1.0-sources.jar ./target/it/reproducible/target/repo/org/apache/maven/its/reproducible/1.0/reproducible-1.0-sources.jar
--- /tmp/reproducible-1.0-sources.jar
+++ ./target/it/reproducible/target/repo/org/apache/maven/its/reproducible/1.0/reproducible-1.0-sources.jar
├── zipinfo {}
│ @@ -1,10 +1,10 @@
│  Zip file size: 7016 bytes, number of entries: 27
│ --rw-r--r--  2.0 unx       25 b- defN 19-Aug-21 18:28 META-INF/MANIFEST.MF
│  drwxr-xr-x  2.0 unx        0 b- stor 19-Aug-21 18:28 META-INF/
│ +-rw-r--r--  2.0 unx       25 b- defN 19-Aug-21 18:28 META-INF/MANIFEST.MF
│  drwxr-xr-x  2.0 unx        0 b- stor 19-Aug-21 18:28 dir-A/
│  drwxr-xr-x  2.0 unx        0 b- stor 19-Aug-21 18:28 dir-C/
│  drwxr-xr-x  2.0 unx        0 b- stor 19-Aug-21 18:28 dir-b/
│  drwxr-xr-x  2.0 unx        0 b- stor 19-Aug-21 18:28 dir-b/B2/
│  drwxr-xr-x  2.0 unx        0 b- stor 19-Aug-21 18:28 dir-b/B4/
│  drwxr-xr-x  2.0 unx        0 b- stor 19-Aug-21 18:28 dir-b/b1/
│  drwxr-xr-x  2.0 unx        0 b- stor 19-Aug-21 18:28 dir-b/b3/
├── zipnote «TEMP»/diffoscope_228psws7_10/tmptcizfwbk_.zip
│ @@ -1,11 +1,11 @@
│ -Filename: META-INF/MANIFEST.MF
│ +Filename: META-INF/
│  Comment:
│
│ -Filename: META-INF/
│ +Filename: META-INF/MANIFEST.MF
│  Comment:
│
│  Filename: dir-A/
│  Comment:
│
│  Filename: dir-C/
│  Comment:
├── filetype from file(1)
│ @@ -1 +1 @@
│ -Zip archive data, at least v2.0 to extract, compression method=deflate
│ +Zip archive data, at least v1.0 to extract, compression method=store
```

@jorsol Would you mind to create a PR this. Your analysis is correct.

michael-o commented:
I will assign issues in JIRA to me and handle this PR. Muchas gracias por tu ayuda!
